File: langs/StdLibs/one.swift
import Foundation

public enum OneError : Error {
    case RuntimeError(String)
}

class OneRegex { 
    class func matchFromIndex(pattern: String, input: String, offset: Int) -> [String]? {
        let regex = try! NSRegularExpression(pattern: "^" + pattern)
        guard let match = regex.firstMatch(in: input, range: NSMakeRange(offset, input.utf16.count - offset)) else { return nil }
        
        var results = [String]()
        for i in 0 ... match.numberOfRanges - 1 {
            let range = match.range(at: i)
            let groupValue = input.substring(with: Range(range, in: input)!)
            results.append(groupValue)
        }        
        return results        
    }
}

class OneReflect {
    static var classes = [String: OneClass]()

    class func getClass(obj: Any?) -> OneClass? {
        let className = String(describing: type(of: obj!))
        return getClassByName(name: className)
    }

    class func getClassByName(name: String) -> OneClass? {
        return classes[nameKey(name)]
    }

    class func addClass(_ cls: OneClass) -> OneClass {
        classes[nameKey(cls.name)] = cls
        return cls
    }

    class func nameKey(_ name: String) -> String {
        return name.replacingOccurrences(of: "_", with: "").lowercased()
    }
}

class OneClass {
    var name: String
    var fields = [String: OneField]()
    var methods = [String: OneMethod]()

    init(name: String) {
        self.name = name
    }

    func getField(name: String) -> OneField? {
        return fields[OneReflect.nameKey(name)]
    }

    func getMethod(name: String) -> OneMethod? {
        return methods[OneReflect.nameKey(name)]
    }

    func addField(_ field: OneField) -> OneClass {
        fields[OneReflect.nameKey(field.name)] = field;
        return self;
    }

    func addMethod(_ method: OneMethod) -> OneClass {
        methods[OneReflect.nameKey(method.name)] = method;
        return self;
    }
}

class OneField {
    var name: String
    var isStatic: Bool
    var type: String
    var getter: (_ obj: Any?) -> Any?
    var setter: (_ obj: Any?, _ value: Any?) -> Void

    init(_ name: String, _ isStatic: Bool, _ type: String, _ getter: @escaping (_ obj: Any?) -> Any?, _ setter: @escaping (_ obj: Any?, _ value: Any?) -> Void) {
        self.name = name
        self.isStatic = isStatic
        self.type = type
        self.getter = getter
        self.setter = setter
    }    

    func getValue(obj: Any?) -> Any? {
        return getter(obj)
    }

    func setValue(obj: Any?, value: Any?) {
        setter(obj, value)
    }

}
class OneMethodArgument {
    var name: String
    var type: String

    init(_ name: String, _ type: String) {
        self.name = name
        self.type = type
    }    
}

class OneMethod {
    var name: String
    var isStatic: Bool
    var returnType: String
    var args: [OneMethodArgument]
    var invoker: (_ obj: Any?, _ args: [Any]) -> Any?

    init(_ name: String, _ isStatic: Bool, _ returnType: String, _ args: [OneMethodArgument], _ invoker: @escaping (_ obj: Any?, _ args: [Any]) -> Any?) {
        self.name = name
        self.isStatic = isStatic
        self.returnType = returnType
        self.args = args
        self.invoker = invoker
    }

    func call(obj: Any?, args: [Any]) -> Any? {
        return invoker(obj, args)
    }
}

File: src/StdLib/one.ts
export class Regex {
    static matchFromIndex(pattern: string, input: string, offset: number) {
        const regex = new RegExp(pattern, "gy");
        regex.lastIndex = offset;
        const matches = regex.exec(input);
        return matches === null ? null : Array.from(matches);
    }
}

export class Reflect {
    static classes: { [name: string]: Class } = {};
    
    static getClass(obj: any): Class { return this.classes[obj.constructor.name.toLowerCase()]; }
    static getClassByName(name: string): Class { return this.classes[name.toLowerCase()]; }

    static setupClass(cls: Class) { this.classes[cls.name.toLowerCase()] = cls; }
}

export class Class {
    name: string;

    fields: { [name: string]: Field; } = {};
    methods: { [name: string]: Method; } = {};

    constructor(public typeObj: any, fields: Field[], methods: Method[]) {
        this.name = typeObj.name;

        for (const field of fields) {
            field.cls = this;
            this.fields[field.name.toLowerCase()] = field;
        }

        for (const method of methods) {
            method.cls = this;
            this.methods[method.name.toLowerCase()] = method;
        }
    }

    getField(name: string) { return this.fields[name.toLowerCase()]; }
    getMethod(name: string) { return this.methods[name.toLowerCase()]; }

    getFields() { return Object.values(this.fields); }
    getMethods() { return Object.values(this.methods); }
}

export class Field {
    cls: Class;

    constructor(public name: string, public isStatic: boolean, public type: string) { }

    getValue(obj: any) {
        const realObj = this.isStatic ? this.cls.typeObj : obj;
        return realObj[this.name];
    }

    setValue(obj: any, value: any){
        const realObj = this.isStatic ? this.cls.typeObj : obj;
        realObj[this.name] = value;
    }
}

export class Method {
    cls: Class;

    constructor(public name: string, public isStatic: boolean, public returnType: string, public args: MethodArgument[]) { }

    call(obj: any, args: any[]) {
        if (args.length !== this.args.length)
            throw new Error(`Expected ${this.args.length} arguments, but got ${args.length} in ${this.cls.name}::${this.name} call!`);

        const realObj = this.isStatic ? this.cls.typeObj : obj;
        const method = realObj[this.name]
        return method.apply(obj, args);
    }
}

export class MethodArgument {
    constructor(public name: string, public type: string) { }
}
